### PR TITLE
Parallel support for NumberArray

### PR DESCRIPTION
**Pull request by roystgnr**

### CODE REVIEW

File: src/numerics/include/metaphysicl/numberarray.h [hunk 1]
```diff
@@ -130,6 +130,9 @@ class NumberArray
   typename entry_type<i>::type& insert()
     { return _data[i]; }
 
+  T & insert(std::size_t i)
+    { return _data[i]; }
```
Comment (lindsayad) [inline src/numerics/include/metaphysicl/numberarray.h:134]: bounds-check assertion?

File: src/numerics/include/metaphysicl/parallel_numberarray.h [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        // resize the structure type to account for padding, if any
+        timpi_call_mpi
+          (MPI_Type_create_resized (tmptype, 0, sizeof(MetaPhysicL::NumberArray<N,T>),
+                                    &static_type));
+
+        timpi_call_mpi
+          (MPI_Type_free (&tmptype));
+
+        SemiPermanent::add
+          (std::make_unique<ManageType>(static_type));
```
Comment (lindsayad) [inline src/numerics/include/metaphysicl/parallel_numberarray.h:83]: This is the code that ensures we don't leak, right?

Comment (roystgnr) [reply src/numerics/include/metaphysicl/parallel_numberarray.h:83]: Right.

File: src/numerics/include/metaphysicl/numberarray.h (hunk 1 above)
Comment (roystgnr) [reply src/numerics/include/metaphysicl/numberarray.h:134]: I am being lazy there.  I'll add one.

Comment (roystgnr) [reply src/numerics/include/metaphysicl/numberarray.h]: Done, and I went through the others while I was at it.

Even admitting this wasn't originally intended as a public project, it's painfully embarrassing to see that that header didn't have a single assertion yet.

Comment (lindsayad) [reply src/numerics/include/metaphysicl/numberarray.h]: Haha. I think you can see you were mimicking the underlying data container as much as possible :smile: No protections on a C array!

**Comment by lindsayad** [comment]
That min gcc failure seems consistent

**Comment by roystgnr** [comment]
If only it were.  The first was a parallel failure in a new test; this one's serial in an old test.

**Comment by lindsayad** [comment]
Lol

**Comment by roystgnr** [comment]
This is super confusing.  I get a warning from g++ 11.2 that certainly looks relevant to the failing test on min gcc:

```
../../src/numerics/include/metaphysicl/dualnumber.h:156:5: warning: ‘<anonymous>’ may be used uninitialized [-Wmaybe-uninitialized]
  156 |     _val(dns.value())
      |     ^~~~~~~~~~~~~~~~~
../../test/nd_derivs_unit.C:312:61: note: ‘<anonymous>’ was declared here
  312 |   DualNumberSurrogate<double, NumberArray<2, double*>> dns(0);
```

But I can't replicate the error and I don't understand the problem and I definitely don't understand how anything I did might have affected it.

**Comment by lindsayad** [comment]
I don't understand why it would warn about `value()` though since `_value` AFAICT is clearly initialized in all the constructors

**Comment by lindsayad** [comment]
If there was actual failure to initialize, I would expect `valgrind` to error, but that test runs valgrind clean for me

**Comment by roystgnr** [comment]
I can't replicate this at all on my workstation (no gcc old enough, may be the problem), but I can on rod, so that's a start.  Did you try valgrind on a g++-7 build?

**Comment by roystgnr** [comment]
Okay, this is just horrifying.  With the same modules as `Min gcc`, from my build subdirectory on rod,

```
../configure && make clean && make -j 20 -l 60 && make -j 20 -l 60 check
```
reliably fails, but if I add `CXXFLAGS=''` to override the default `-g -O2 -Wall` then it reliably passes.  If I specify `-O2` or `-O3` manually, failure; `-O1` or `-O0`, success.

If it wasn't for the warning from gcc 11 I'd swear this was a compiler bug.

**Comment by lindsayad** [comment]
So I also get the gcc 11 warning when I compile with -O2 ... but somewhat analogously to min gcc (but to a far less horrifying degree) I *do not* get a compiler warning with -O0 with gcc 11

**Comment by lindsayad** [comment]
This reminds me of https://github.com/cxong/tinydir/issues/72

**Comment by roystgnr** [comment]
Even `-Os` fails for me.  But that's weak enough to let me get at this stuff in gdb.

And ... is it me, or is `DualNumberSurrogate(T &&)` inherently screwed up?  We're taking a reference in _value, and if we're taking it from an rvalue reference that's a pretty big hint that it might be a dangling reference very shortly, right?  I'm not sure why g++-7 is deciding to make a temporary out of a literal "0" but I can't swear that they're definitely *wrong* to do so.

And if I remove that constructor then the bug goes away for me.  Let me see if CI agrees.

**Comment by roystgnr** [comment]
In fact, I'm not even sure that goes far enough.  Isn't taking a pointer or a non-const reference to a literal supposed to be forbidden?  I'm not sure why the compilers are allowing it for us here.  I'm going to just use that named `zero` variable in this case too, instead of `0`, lest some future compiler complain, just as soon as I see whether this first change passes or fails.

**Comment by lindsayad** [comment]
Oh wow yea that's terrible

**Review by lindsayad — APPROVED** [review]

**Comment by roystgnr** [comment]
Well, this is fun.  With [53ac534](https://github.com/libMesh/MetaPhysicL/commit/53ac534babf6e86d2814b0ca6fdf9e14791c0c28) plus a couple cherry-picked commits to let me put `enableFPE(true)` in `derivs_unit.C`, I get an FPE from the test of `random_quant - tan(atan(random_quant))` ... with clang-9 -O2 on my workstation, anyway.  I don't seem to be able to get that from later clang compilers, or without optimization, or from gcc, or in CI, or if I put that `atan()` output into a named rather than a temporary variable.

**Comment by roystgnr** [comment]
This is pure madness.  If I check out c8091f4b86e07b42d222c2832bc67c57fe0df8e5, then cherry-pick f1aeaa5f61dde7e90ced6f7e1d8f2cef6257cf21 a271ee25e8c517816d55857ac8a8fccd0a0281ff, then modify `derivs_unit.C` to `enableFPE(true)` and to skip the `test_func_derivatives` sections, then the result passes `make check` after a `../configure CXX=clang++-9 CC=clang-9 CXXFLAGS='-O2 -g'` build.

If I then cherry-pick 1b17c9f9344a5a5b728f65677da00d206f105fb4, the result fails with the FPE above.

There is nothing in that failing commit except some asserts and `static_assert`s.  If I add `-DNDEBUG` to my `$CXXFLAGS` to disable the runtime asserts, `make check` passes again.

I'm using clang "version 9.0.1-16.1ubuntu1", but I don't see anything newer (except for 10+, of course) on their releases page.

**Comment by lindsayad** [comment]
Sounds wild. Is the test valgrind clean?

**Comment by lindsayad** [comment]
I don't think so

**Comment by roystgnr** [comment]
I can boil derivs_unit.C down to a single `foo - atan(tan(foo))` of `DN<float,NA<10,float>>` and get the same failure ... the more I dig into this, the more this feels like there has to be a dangling reference somewhere, I just have no idea where.  Cherry-picking [2757709](https://github.com/libMesh/MetaPhysicL/pull/16/commits/27577094c0e7cc3a621bbffb77d17705999944a9) doesn't seem to help this time.

**Comment by roystgnr** [comment]
It even starts passing if I try to use `N = 7` or less.  This sure feels like we have to be smashing something on the stack somehow.  But I can try `N = 800` and neither gcc 11 nor clang 13 hit any errors.

**Comment by lindsayad** [comment]
Do you want me to take a shot with memory sanitizer?

**Comment by roystgnr** [comment]
That would be quite welcome.  I'd love to know if anyone else can even replicate this, regardless of whether anyone else can debug it.

**Comment by roystgnr** [comment]
I have a failing case now where in the middle I do:
```
  const float & f = in.value();
//  std::cerr << "f is " << f << std::endl;
  std::cos(f);
```

If I uncomment the print, I see "f is 0" and the test case completes successfully.
If I leave the print commented out, I get SIGFPE later in the code.
If I comment out the call to **cos**, the test case completes successfully.

I'm going to bed.  This is now pure **distilled** madness.

**Comment by roystgnr** [comment]
Just not giving clang any rvalue references to deal with seems to fix the problem.  I think I'm going to call that our workaround and not spend any more time on this.

If I was working with a heap-allocated class I'd still worry about undefined behavior and dangling references, but this is all with just `DualNumber<float, NumberArray<N,float>>`.  Feels like it's got to be a compiler bug.

**Comment by lindsayad** [comment]
I don't see any memory failures from `derivs_unit` though :-/

**Comment by lindsayad** [comment]
actually that trace is pretty useless. I don't even know how there could be uninitialized bytes on that line

**Comment by lindsayad** [comment]
I'm hitting this:
```
../../src/utilities/src/metaphysicl_exceptions.C:34: error: "_GNU_SOURCE" redefined [-Werror]
```

**Pull request merged**